### PR TITLE
fix various path-related functions on non-String strings, add tests

### LINKED ISSUE #17024: `splitdrive` of Windows doesn't work with `AbstractString`
It is noted in RCall /Julia 0.5 appveyor build: https://ci.appveyor.com/project/simonbyrne/rcall-jl/build/1.0.185/job/lu63f6trv0xtdk6o#L232

```
ERROR: LoadError: LoadError: ArgumentError: regex matching is only available for the String type; use String(s) to convert
 in splitdrive at .\path.jl:38 [inlined]
 in joinpath(::UTF16String, ::String) at .\path.jl:81
 in joinpath(::UTF16String, ::String, ::String, ::Vararg{String,N}) at .\path.jl:77
 in include_from_node1(::String) at .\loading.jl:426 (repeats 2 times)
 in macro expansion; at .\none:2 [inlined]
```

It is likely caused by the conflict between `splitdrive` and `match`. `splitdrive` allows `path` to be `AbstractString` [(link)](https://github.com/JuliaLang/julia/blob/00ba80d22760c60bd87c12ce865ff6ace544594e/base/path.jl#L38), but `match` only takes `String` type [(link)](https://github.com/JuliaLang/julia/blob/00ba80d22760c60bd87c12ce865ff6ace544594e/base/regex.jl#L174).

### PR DESCRIPTION
**Pull request by StefanKarpinski**
fixes #17024; discovered some buggy corner cases for splitext

### CODE REVIEW

File: base/path.jl [hunk 1]
```diff
@@ -122,17 +122,17 @@ function normpath(path::AbstractString)
...
 
-abspath(a::AbstractString) = normpath(isabspath(a) ? a : joinpath(pwd(),a))
+abspath(a::String) = normpath(isabspath(a) ? a : joinpath(pwd(),a))
 abspath(a::AbstractString, b::AbstractString...) = abspath(joinpath(a,b...))
 
 if is_windows()
 function realpath(path::AbstractString)
-    path = cwstring(path)
-    buf = zeros(UInt16, length(path))
+    p = cwstring(path)
```
Comment (tkelman) [inline base/path.jl]: was there a bug here or is this just an annoying type-stability thing?

**Comment by tkelman** [comment]
was the original bug due to trying to call these on mixed string types? don't all the tests here only call them with all inputs the same type?

File: base/path.jl (hunk 1 above)
Comment (StefanKarpinski) [inline base/path.jl]: Just avoiding the type instability.

**Comment by StefanKarpinski** [comment]
The original bug wasn't because of mixed types, afaict, it was because splitpath didn't ensure conversion to `String` before applying regex matching, which therefore failed. I could modify this to test different combinations, but it seems like overkill.

**Comment by StefanKarpinski** [comment]
Should probably be backported to 0.4 with `s/\bString\b/ASCIIString/g` applied.

**Comment by tkelman** [comment]
on second thought paths can be unicode on most recent filesystems, so wouldn't want to restrict these signatures to ascii-only or throw if the converting fallbacks don't work, right?

**Comment by StefanKarpinski** [comment]
True, probably better to backport with String replaced by ByteString?

**Comment by tkelman** [comment]
I'm leaning more towards not touching it, has anyone encountered the bug in real usage of 0.4?

**Comment by StefanKarpinski** [comment]
True – given the lack of complaints, let's leave it.

**Pull request merged**